app/controllers/links_controller.rb

```diff
@@ -236,7 +259,26 @@ def set_link
     if params[:id].match? /\D+/
       @link = Link.find_by(custom_url: params[:id])
     else
-      @link = Link.find(params[:id])
+      @link = Link.find_by(id: params[:id])
+    end
+  end
+
+  def require_link
+    if @link.nil?
+      raise ActionController::RoutingError.new('Not Found')
```

Comment:
let me know if this is shit; this was just the most "obvious" way i could see to do a 404 without tethering it directly to the setter method itself